### PR TITLE
learn2016: URL parameters can override default initial filters

### PR DESCRIPTION
**Pull request by breville**
The parameters combine filter group name and filter entry name to specify which filters should be checked.

e.g. /learn?activity_type=lesson-plan&length=1hour&length=few-hours

There must be at least one valid filterGroupName and filterEntryName combination to override defaults.  The list is whitelisted against the real list of filters.  Activity type can't be overridden on the /learn/robotics page.

### CODE REVIEW

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 1]
```diff
@@ -423,9 +423,52 @@ function getFilters({robotics, mobile}) {
   return {filters, initialFilters, hideFilters};
 }
 
+function getUrlParameters(filters) {
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: Would be great to have an @param comment here, and also an @returns

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 2]
```diff
@@ -423,9 +423,52 @@ function getFilters({robotics, mobile}) {
...
+  let providedParameters = null;
+
+  const parameters = decodeURIComponent(window.location.search.substring(1)).split('&');
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: Elsewhere we use query-string to access query params (https://github.com/sindresorhus/query-string). Would be good to use the same thing here.

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 3]
```diff
@@ -423,42 +422,33 @@ function getFilters({robotics, mobile}) {
...
+
+  let parameters = queryString.parse(location.search);
+
+  for (const name in parameters) {
+    if (typeof parameters[name] === "string") {
+      // Convert items with a single filter entry into an array containing that
+      // string.
+      parametersObject[name] = [parameters[name]];
+    } else {
+      parametersObject[name] = parameters[name];
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: I think you could do this as a one liner: `parametersObject[name] = [].concat(parameters[name])`

It's possible that doing so sacrifices some amount of understandability in order to gain brevity. Up to you which you prefer.

Comment (breville) [reply apps/src/tutorialExplorer/tutorialExplorer.js]: Cool technique, though I had to run it manually to convince myself it worked, so I think I will keep the expansive version for now :)

**Comment by tanyaparker** [comment]
When I test latest state of staging locally, it still has "Beginner" selected by default. On this branch it seems to have lost those defaults?

**Comment by breville** [comment]
@tanyaparker and @Bjvanminnen I made some fixes, notably:
- /learn/robotics now forces activity type of "robotics", can't be overridden by URL.
- default initialFilters, when no URL overrides, work again.
- redirects on hourofcode.com now include a reconstructed query string.
  e.g. "hourofcode.com/learn?param=true" now redirects to "hourofcode.com/us/learn?param=true" whereas before the query string would be removed, and so it wouldn't be useful as a publishable URL.

File: apps/src/tutorialExplorer/tutorialExplorer.js [hunk 4]
```diff
@@ -508,9 +509,48 @@ function getFilters({robotics, mobile}) {
...
+/*
+ * Parse URL parameters to retrieve an override of initialFilters.
+ *
+ * @return {object} - Returns an object containing arrays of strings.  Each
+ *   array is named for a filterGroup name, and each string inside is named
+ *   for a filter entry.  Note that this is not currently white-listed against
+ *   our known name of filterGroups/entries, but invalid entries should be
+ *   ignored in the filtering user experience.
+ */
+function getUrlParameters(robotics) {
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/tutorialExplorer.js]: robotics is a boolean? i'd add a param comment

**Comment by breville** [comment]
@tanyaparker and @Bjvanminnen I revised this functionality again so that it only replaces default filters if it gets at least one valid filterGroupName+filterEntryName in the URL, and it effectively whitelists them all against the real list of filters.

**Pull request merged**